### PR TITLE
Handle cases where selection API doesn't apply

### PR DESCRIPTION
**Pull request by jonleighton**
The selection API only applies to certain <input> types:

https://html.spec.whatwg.org/multipage/#do-not-apply

This commit ensures that we handle that correctly.

Some notes:

1. TextControl::set_dom_selection_direction now calls
   set_selection_range(), which means that setting selectionDirection will
   now fire a selection event, as it should per the spec.

2. There is a test for the firing of the select event in
   tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html,
   however the test did not run due to this syntax error:

   (pid:26017) "ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/semantics/forms/textfieldselection/select-event.html:50:11 missing = in const declaration"

   This happens due to the us of the "for (const foo of ...)" construct.
   Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
   this should actually work, so it's somewhat unsatisfying to have to
   change the test.

3. I removed tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html
   because it doesn't seem to add any extra value - the selection API
   always applies to textarea elements, and the API is tested elsewhere.

4. If an `<input>`'s type is unset, it defaults to a text, and the
   selection API applies. Also, if an `<input>`'s type is set to an
   invalid value, it defaults to a text too. This second case doesn't
   currently work, and I'll need to do more restructuring of the code in
   a future commit. See discussion with nox in IRC:
   https://mozilla.logbot.info/servo/20171201#c13946454-c13946594

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19461)

### CODE REVIEW

**Comment by nox** [comment]
Please remove my name from the commit message, I get a notification every time it's pushed on GH.

**Comment by jonleighton** [comment]
Sorry about that, done. Do you want to still be assigned? Not sure if you want to be involved with this, @KiChjang has reviewed several of my previous patches but this obviously relates somewhat to what we discussed in IRC.

(Also not sure if you got assigned because I mentioned you or whether it was just luck of the draw, I don't really know how the bot works, but anyway, feel free to tell me what you prefer.)

**Comment by jonleighton** [comment]
I've now submitted https://github.com/servo/servo/pull/19471 which cleans up the handling of the type attribute. If that gets merged, I can make this change a bit cleaner and the tests for `<input type=invalidvalue>` will pass.

File: components/script/dom/htmlinputelement.rs [hunk 1]
```diff
@@ -408,6 +408,17 @@ impl TextControl for HTMLInputElement {
     fn textinput(&self) -> &DomRefCell<TextInput<ScriptToConstellationChan>> {
         &self.textinput
     }
+
+    fn selection_api_applies(&self) -> bool {
```
Comment (KiChjang) [inline components/script/dom/htmlinputelement.rs]: Can you link to the relevant section of the spec that defines this behaviour?

File: tests/wpt/metadata/html/semantics/forms/textfieldselection/select-event.html.ini [hunk 2]
```diff
@@ -1,23 +1,146 @@
...
+
+  [input type password: selectionDirection a second time (must not fire select)]
+    expected: FAIL
+
+  [input type password: setSelectionRange() a second time (must not fire select)]
+    expected: FAIL
+
+  [input type password: setRangeText()]
+    expected: FAIL
+
```
Comment (KiChjang) [inline tests/wpt/metadata/html/semantics/forms/textfieldselection/select-event.html.ini]: Why are there so many expected test failures?

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html [hunk 3]
```diff
@@ -47,10 +47,10 @@
   }
 ];
 
-for (const el of els) {
+els.forEach((el) => {
   const elLabel = el.localName === "textarea" ? "textarea" : "input type " + el.type;
 
-  for (const action of actions) {
+  actions.forEach((action) => {
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Why the change?

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html [hunk 4]
```diff
@@ -1,29 +0,0 @@
-<!DOCTYPE html>
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html]: Why is this test file removed?

**Review by KiChjang — COMMENTED** [review]
Great work! I don't see much problems with the code, but I have some comments about the tests.

File: components/script/dom/htmlinputelement.rs (hunk 1 above)
Comment (jonleighton) [reply components/script/dom/htmlinputelement.rs]: Yep

File: tests/wpt/metadata/html/semantics/forms/textfieldselection/select-event.html.ini (hunk 2 above)
Comment (jonleighton) [reply tests/wpt/metadata/html/semantics/forms/textfieldselection/select-event.html.ini]: These are just things that haven't yet been implemented:

* `select()` method
* `setRangeText()` method
* Code to avoid firing the `select` event when the selection hasn't actually moved

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html (hunk 3 above)
Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: This is explained in the commit message:

> There is a test for the firing of the select event in
tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html,
however the test did not run due to this syntax error:
>
> (pid:26017) "ERROR:script::dom::bindings::error: Error at http://web-platform.test:8000/html/semantics/forms/textfieldselection/select-event.html:50:11 missing = in const declaration"
>
> This happens due to the us of the "for (const foo of ...)" construct.
> Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
this should actually work, so it's somewhat unsatisfying to have to
change the test.

Definitely not ideal to have to change this, so I'm open to other suggestions! But without changing this, the changes here are not fully tested (specifically, setting `selectionDirection` will now call `set_selection_range` and fire the `select` event, and this test file verifies that behaviour).

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html (hunk 4 above)
Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html]: Also explained in the commit message:

> I removed tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html
because it doesn't seem to add any extra value - the selection API
always applies to textarea elements, and the API is tested elsewhere.

Comment (KiChjang) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html]: I'd rather not remove this haphazardly without contacting the authors or upstream W3C about this. Note that this is not a servo-specific test - any changes you make under `tests/wpt/web-platform-tests` are shared across all browsers.

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html (hunk 3 above)
Comment (KiChjang) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Does `let` instead of `const` work? If not, then this looks like it's a problem with our implementation, and not the tests.

Comment (jdm) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Our spidermonkey version doesn't support `for (const foo of bar)`. We modify tests to use `for (let foo of bar)` instead when we want to be able to run them.

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: I think I tried `let` but was having trouble for some reason. I'll have another go.

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html (hunk 4 above)
Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application-textarea.html]: Fair enough, I'll reinstate it.

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html (hunk 3 above)
Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Right, with `let`, when the function gets executed `action` is always set to the last item in `actions`, which results in a call to `setRangeText`, which fails. This seems wrong: `let` seems to be behaving like `var`. Is this another difference in Servo's version of SpiderMonkey?

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application.html [hunk 5]
```diff
@@ -7,45 +7,90 @@
 <script src="/resources/testharnessreport.js"></script>
 <div id="log"></div>
 <script>
-  var types = ["hidden", "email", "datetime-local", "date", "month", "week", "time", "number", "range", "color", "checkbox", "radio", "file", "submit", "image", "reset", "button"]; //types for which the API doesn't apply
-  var types2 = ["text", "search", "tel", "url", "password"]; //types for which the API applies
+  var nonApplicableTypes = ["hidden", "email", "datetime-local", "date", "month", "week", "time", "number", "range", "color", "checkbox", "radio", "file", "submit", "image", "reset", "button"];
+  var applicableTypes = ["text", "search", "tel", "url", "password", "aninvalidtype", null];
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application.html]: Why are `null` and `"aninvalidtype"` accepted? Is it because they're both going to be treated as `"text"`?

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/selection-not-application.html]: Correct. My first implementation only worked with the former, but now since I've done the `InputType` refactoring, both are correctly treated as `"text"`.

File: tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html (hunk 3 above)
Comment (jdm) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Not one that I'm aware of! Want to make an isolated example demonstrating it and file a new issue?

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Sure: https://github.com/servo/servo/issues/19535

How should we proceed with this PR?

Comment (KiChjang) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: My inclination would be to revert this change and mark them as expected failures.

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Or maybe put the modified version in `tests/wpt/mozilla/` for the time being? Seems a shame not to have it tested in any way, as then we won't be sure we're not introducing regressions.

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: Though as far as I can see those tests (in `tests/wpt/mozilla/`) don't actually get run by the CI? So maybe that wouldn't particularly help.

Comment (jdm) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: No, tests/wpt/mozilla/ definitely get run by CI. Adding a modified version to that directory is a good idea.

Comment (nox) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: > 15:24 &lt;annevk> nox: I think it’s considered acceptable to rewrite if not too much of a concession in readability

We could just use the `forEach` method in that test.

Comment (jonleighton) [reply tests/wpt/web-platform-tests/html/semantics/forms/textfieldselection/select-event.html]: @jdm are you happy to do as @nox suggests (if so, this can be merged as it is), or shall I add the modified test in tests/wpt/mozilla?

**Comment by KiChjang** [comment]
@jonleighton This should be good enough; thanks for your work!

@bors-servo r+

**Comment by jonleighton** [comment]
Hooray, thanks ✨
